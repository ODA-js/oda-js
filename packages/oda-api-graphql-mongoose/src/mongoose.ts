import { fromGlobalId } from 'oda-isomorfic';
import { set, get } from 'lodash';
import getLogger from 'oda-logger';
let logger = getLogger('oda-api-graphql:api-mongoose');

import {
  ConnectorsApiBase,
  listIterator,
  SecurityContext,
  consts,
  pagination,
  detectCursorDirection,
  Filter as FilterBase,
  RegisterConnectorsBase,
} from 'oda-api-graphql';

const { forward } = listIterator;
const { DIRECTION } = consts;
const { Filter } = FilterBase;

function unfoldQuery(
  obj: any | any[],
  operations: object,
  parent: string[],
  res?: object,
) {
  logger.trace('unfoldQuery');
  if (typeof obj === 'object') {
    if (Array.isArray(obj)) {
      return obj.map(item => unfoldQuery(item, operations, parent));
    } else {
      if (!res) {
        res = {};
      }
      Object.keys(obj).forEach(key => {
        let setKey = key;
        let index = setKey.match(/^at_(\d*)$/);
        if (index && index[1]) {
          setKey = index[1];
        }
        if (operations.hasOwnProperty(key)) {
          if (operations[key] === 'array') {
            res[key] = unfoldQuery(obj[key], operations, parent);
          } else {
            res[parent.join('.')] = obj;
          }
        } else {
          unfoldQuery(obj[key], operations, [...parent, setKey], res);
        }
      });
      return res;
    }
  } else {
    return obj;
  }
}

export class FilterMongoose {
  public static parse(node, idMap = { id: '_id' }, id: boolean = false) {
    logger.trace('parse %o', node);
    let result = Filter.parse(node, idMap, id);
    if (typeof result === 'object') {
      return unfoldQuery(result, Filter.types, []);
    }
    return result;
  }
}

export default class MongooseApi<
  RegisterConnectors extends RegisterConnectorsBase,
  Payload extends object
> extends ConnectorsApiBase<RegisterConnectors, Payload> {
  public mongoose: any;

  constructor({
    mongoose,
    connectors,
    name,
    securityContext,
  }: {
    mongoose: any;
    name: string;
    connectors: RegisterConnectors;
    securityContext: SecurityContext<RegisterConnectors>;
  }) {
    super({ connectors, securityContext, name });
    this.mongoose = mongoose;
  }

  protected initSchema(name, schema) {
    logger.trace('%s unfoldQuery', name);
    this.schema = schema;
    if (!this.mongoose.models[name]) {
      this.model = this.mongoose.model(name, schema);
    } else {
      this.model = this.mongoose.model(name);
    }
  }
  public async getCount(args) {
    logger.trace('getCount %o', args);
    let query = this.getFilter(args);
    if (query.hasOwnProperty().length > 0) {
      logger.trace('getCount use countDocuments %o', query);
      return await this.model.countDocuments(query);
    } else {
      logger.trace('getCount use estimatedDocumentCount');
      return await this.model.estimatedDocumentCount();
    }
  }

  public getFilter(args: { filter: any; idMap: any }) {
    logger.trace('getFilter %o', args);
    if (args.filter) {
      return FilterMongoose.parse(args.filter, args.idMap);
    } else {
      return {};
    }
  }

  public toJSON(obj): Payload {
    logger.trace('%s toJSON', this.name);
    return super.toJSON(obj && obj.toObject ? obj.toObject() : obj);
  }

  public ensureId(obj) {
    logger.trace('%s ensureId', this.name);
    if (obj) {
      return this.toJSON({
        ...obj,
        id: obj._id,
      } as Payload);
    } else {
      return obj as Payload;
    }
  }

  protected async _getList(args, checkExtraCriteria?) {
    logger.trace('%s _getList', this.name);
    let hasExtraCondition = typeof checkExtraCriteria !== 'undefined';
    let query: any = this.getFilter(args);
    let sort = detectCursorDirection(args);
    let cursor = pagination(args);

    let result = [];

    let move: any = {};

    if (cursor.after || cursor.before) {
      const detect = (name: string, value: any) => ({
        [name]:
          sort[name] === DIRECTION.BACKWARD
            ? { [cursor.after ? '$lt' : '$gt']: value }
            : { [cursor.before ? '$lt' : '$gt']: value },
      });
      let sortKeys = Object.keys(sort);
      if (sortKeys.length > 1) {
        let current = await this.findOneById(cursor.after || cursor.before);
        let find = sortKeys.filter(f => f !== '_id');
        find.push('_id');
        const or = [];
        while (find.length > 0) {
          const len = find.length;
          or.push(
            find.reduce((prev, f, index) => {
              const curr =
                index === len - 1
                  ? detect(f, current[f])
                  : {
                      [f]: { $eq: current[f] },
                    };
              prev = {
                ...prev,
                ...curr,
              };
              return prev;
            }, {}),
          );
          find.pop();
        }
        move = { $or: or };
      } else {
        move = {
          _id: {
            [sort._id === DIRECTION.FORWARD ? '$gt' : '$lt']:
              cursor.after || cursor.before,
          },
        };
      }
    }

    if (Object.keys(query).length > 0) {
      if (Object.keys(move).length > 0) {
        query = {
          $and: [move, query],
        };
      }
    } else {
      if (Object.keys(move).length > 0) {
        query = move;
      }
    }

    let pageSize = 10;
    let iterator = forward(async (step: number) => {
      logger.trace('%s _getList:forward', this.name);

      return await new Promise<any[]>((res, rej) => {
        let rz = this.model
          .find(query)
          .sort(sort)
          .skip(cursor.skip + step * pageSize)
          .limit(pageSize);

        if (
          this.connectors &&
          this.connectors.transaction &&
          this.connectors.transaction.session &&
          this.connectors.transaction.session.mongoose
        ) {
          rz = rz.session(this.connectors.transaction.session.mongoose);
        }

        rz.exec((err, data) => {
          if (err) {
            logger.error('%s _getList:forward:error', this.name);
            rej(err);
          } else {
            logger.trace('%s _getList:forward:done', this.name);
            res(data);
          }
        });
      });
    }, pageSize);

    for await (let source of iterator) {
      if (
        (cursor.limit && result.length < cursor.limit) ||
        (!cursor.limit || cursor.limit <= 0)
      ) {
        logger.trace('%s _getList:more', this.name);
        if (await this.readSecure(source)) {
          if (hasExtraCondition) {
            if (await checkExtraCriteria(this.toJSON(source))) {
              result.push(source);
            }
          } else {
            result.push(source);
          }
        }
      } else {
        logger.trace('%s _getList:limit', this.name);
        break;
      }
    }

    this.storeToCache(result);
    return result;
  }

  protected async _create(obj: Payload) {
    logger.trace('%s _create', this.name);
    let res = new this.model(obj);
    if (
      this.connectors &&
      this.connectors.transaction &&
      this.connectors.transaction.session &&
      this.connectors.transaction.session.mongoose
    ) {
      res.$session(this.connectors.transaction.session.mongoose);
    } else {
      res.$session(undefined);
    }
    return await res.save();
  }

  protected async _update(record, obj: Payload) {
    logger.trace('%s _update', this.name);
    for (let f in obj) {
      if (obj.hasOwnProperty(f)) {
        record.set(f, obj[f]);
      }
    }
    if (
      this.connectors &&
      this.connectors.transaction &&
      this.connectors.transaction.session &&
      this.connectors.transaction.session.mongoose
    ) {
      record.$session(this.connectors.transaction.session.mongoose);
    } else {
      record.$session(undefined);
    }
    return await record.save();
  }

  protected async _remove(record) {
    logger.trace('%s _remove', this.name);
    if (
      this.connectors &&
      this.connectors.transaction &&
      this.connectors.transaction.session &&
      this.connectors.transaction.session.mongoose
    ) {
      record.$session(this.connectors.transaction.session.mongoose);
    } else {
      record.$session(undefined);
    }
    return await record.remove();
  }

  public async sync({ force = false }: { force?: boolean }) {
    logger.trace('%s sync empty', this.name);
    return;
  }
}
