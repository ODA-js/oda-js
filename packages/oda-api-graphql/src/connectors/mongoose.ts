import { Filter } from '../filter';
import pagination from '../pagination';
import cursorDirection from '../direction';
import { DIRECTION } from '../consts';

import { fromGlobalId } from 'graphql-relay';

import ConnectorsApiBase from './api';

import { forward } from './listIterator';

export default class MongooseApi<RegisterConnectors, Payload> extends ConnectorsApiBase<RegisterConnectors, Payload> {

  public mongoose: any;

  constructor({ mongoose, connectors, user, owner, acls, userGroup }) {
    super({ connectors, user, owner, acls, userGroup });
    this.mongoose = mongoose;
  }

  protected initSchema(name, schema) {
    this.schema = schema;
    if (!this.mongoose.models[name]) {
      // init once
      if (this.user) {
        this.schema.pre('save', this.logUser());
      }
      if (this._viewer) {
        this.schema.pre('save', this.initOwner());
      }
      this.model = this.mongoose.model(name, schema);
    } else {
      this.model = this.mongoose.model(name);
    }
  }

  public async getCount(args) {
    let query = this.getFilter(args);
    return (await this.model.count(query));
  }

  public getFilter(args: { filter: any, idMap: any }) {
    if (args.filter) {
      return Filter.parse(args.filter, args.idMap);
    } else {
      return {};
    }
  };

  public toJSON(obj): Payload {
    return super.toJSON((obj && obj.toObject) ? obj.toObject() : obj);
  }

  public ensureId(obj) {
    if (obj) {
      return this.toJSON({
        ...obj,
        id: obj._id,
      } as Payload);
    } else {
      return obj as Payload;
    }
  }

  protected async _getList(args, checkExtraCriteria?) {
    let hasExtraCondition = typeof checkExtraCriteria !== 'undefined';
    let query: any = this.getFilter(args);
    let sort = cursorDirection(args);
    let cursor = pagination(args);

    let result = [];

    let move: any = {};

    if (cursor.after || cursor.before) {
      const detect = (name: string, value: any) =>
        ({
          [name]: (sort[name] === DIRECTION.BACKWARD) ?
            { [cursor.after ? '$lte' : '$gte']: value } :
            { [cursor.before ? '$lte' : '$gte']: value },
        });
      ;
      let sortKeys = Object.keys(sort);
      if (sortKeys.length > 1) {
        let current = await this.findOneById(cursor.after || cursor.before);
        let find = sortKeys.filter(f => f !== '_id').map(f => detect(f, current[f]));
        find.push({ _id: { $gt: cursor.after || cursor.before } });
        move = find.reduce((prev, curr) => {
          prev = {
            ...prev,
            ...curr,
          };
          return prev;
        }, {});
      } else {
        move = {
          _id: { [sort._id === DIRECTION.FORWARD ? '$gt' : '$lt']: cursor.after || cursor.before },
        };
      }
    }

    if (Object.keys(query).length > 0) {
      if (Object.keys(move).length > 0) {
        query = {
          $and: [
            move,
            query,
          ],
        };
      }
    } else {
      if (Object.keys(move).length > 0) {
        query = move;
      }
    }

    let pageSize = 10;
    let iterator = forward(async (step: number) => {
      return await new Promise<any[]>((res, rej) => {
        this.model.find(query).sort(sort)
          .skip(cursor.skip + step * pageSize)
          .limit(pageSize)
          .exec((err, data) => {
            if (err) {
              rej(err);
            } else {
              res(data);
            }
          });
      });
    }, pageSize);

    for await (let item of iterator) {
      if ((cursor.limit && (result.length < cursor.limit)) || ((!cursor.limit) || (cursor.limit <= 0))) {
        if (this.canView(item)) {
          if (hasExtraCondition) {
            if (await checkExtraCriteria(this.toJSON(item))) {
              result.push(item);
            }
          } else {
            result.push(item);
          }
        }
      } else {
        break;
      }
    }

    this.storeToCache(result);

    return result;
  }

  protected logUser() {
    let _user = () => this.user;
    return function (next) {
      let user = _user();
      if (this.isNew) {
        this.set('createdAt', new Date());
        this.set('createdBy', fromGlobalId(user.id).id);
      } else {
        this.set('updatedAt', new Date());
        this.set('updatedBy', fromGlobalId(user.id).id);
      }
      next();
    };
  }

  protected initOwner() {
    let _owner = () => this._viewer;
    return function (next) {
      if (this.isNew && !this.get('owner')) {
        let owner = _owner();
        if (owner.owner) {
          this.set('owner', owner.owner);
        } else {
          this.set('owner', owner.id);
        }
      }
      next();
    };
  }

  public async sync({ force = false }: { force?: boolean }) { }
}
