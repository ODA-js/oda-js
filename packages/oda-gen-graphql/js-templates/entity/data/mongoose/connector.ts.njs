<#@ context 'entity' -#>
import * as log4js from 'log4js';
let logger = log4js.getLogger('api:connector:#{entity.name}');

import { MongooseApi } from 'oda-api-graphql';
import #{ entity.name }Schema from './schema';
import RegisterConnectors from '../../registerConnectors';
import * as Dataloader from 'dataloader';

import { #{ entity.name } as #{ entity.name }Model } from '../types/model';

export default class #{ entity.name } extends MongooseApi<RegisterConnectors> {
  constructor({mongoose, connectors, user, owner, acls, userGroup}) {
    logger.trace('constructor');
    super({mongoose, connectors, user, acls, userGroup
<#-if( entity.needOwner ){-#>
, owner
<#-} else {-#>
, owner: false
<#-}-#>
    });
    this.initSchema('#{entity.name}', #{ entity.name }Schema());
    this.storeToCache = this.updateLoaders('All Fields');

    this.loaderKeys = {
<#- for (let i = 0, len = entity.loaders.length; i < len; i++) {
    let loaderName = entity.loaders[i];
#>
      by#{loaderName}: '#{entity.unique[i]}',
<#-}#>
<#- entity.complexUniqueIndex.forEach(f=> {
  let findBy = f.fields.map(f=>f.uName).join('And');
  let loadArgs = f.fields.map(f=>`obj.${f.name}`).join(' + ');
    #>
      by#{findBy}: obj => #{loadArgs},
<#});#>
    };

    this.updaters = {
<#- for (let i = 0, len = entity.loaders.length; i < len; i++) {
    let loaderName = entity.loaders[i];
#>
      by#{loaderName}: this.updateLoaders('by#{loaderName}'),
<#-}#>
<#- entity.complexUniqueIndex.forEach(f=> {
  let findBy = f.fields.map(f=>f.uName).join('And');
    #>
      by#{findBy}: this.updateLoaders('by#{findBy}'),
<#});#>
    };

    this.loaders = {
<#- for (let i = 0, len = entity.loaders.length; i < len; i++) {
    let loaderName = entity.loaders[i];
#>
      by#{loaderName}: new Dataloader(keys =>
        Promise.all<#{entity.name}Model>(keys.map(this._findOneBy#{loaderName}.bind(this)) as Promise<#{entity.name}Model>[])
        .then(this.updaters.by#{loaderName})
<#- if(loaderName === 'Id'){#>, {
          cacheKeyFn: key => typeof key !== 'object' ? key : key.toString(),
        }
<#-}-#>),
<#-}#>
<#- entity.complexUniqueIndex.forEach(f=> {
  let findBy = f.fields.map(f=>f.uName).join('And');
  let loadArgs = f.fields.map(f=>`key.${f.name}`).join(' + ');
  let withArgsTypeof = f.fields.map(f=>'${'+`typeof ${f.name}`+'}').join(', ');
    #>
      by#{findBy}: new Dataloader(keys => Promise.all<#{entity.name}Model>(keys.map(this._findOneBy#{findBy}.bind(this)) as Promise<#{entity.name}Model>[])
        .then(this.updaters.by#{findBy}), {
          cacheKeyFn: key => typeof key === 'object' ? (#{loadArgs}) : key.toString(),
        }),
<#});#>
    };
  }

  public async create(payload) {
    logger.trace('create');
    let entity = this.getPayload(payload);
    let result = await  (new (this.model)(entity)).save();
    this.storeToCache([result]);
    return this.ensureId(result ? result.toJSON() : result);
  }
<#- for (let f of entity.args.update.find) {
  let ukey = f.name;
  let type = f.type;
#>

  public async findOneBy#{f.cName}AndUpdate(#{ukey}: #{type}, payload: any) {
    logger.trace(`findOneBy#{f.cName}AndUpdate`);
    let entity = this.getPayload(payload, true);
    let result = await this.loaders.by#{f.cName}.load(#{ukey});
    if(result){
      for (let f in entity) {
        if (entity.hasOwnProperty(f)) {
          result.set(f, entity[f]);
        }
      }
      result = await result.save();
      this.storeToCache([result]);
      return this.ensureId(result ? result.toJSON() : result);
    } else {
      return result;
    }
  }
<#-}#>

<#- entity.complexUniqueIndex.forEach(f=> {
  let findBy = f.fields.map(f=>f.uName).join('And');
  let findArgs = f.fields.map(f=>`${f.name}: ${f.type}`).join(', ');
  let withArgs = f.fields.map(f=>'${'+f.name+'}').join(', ');
  let withArgs2 = f.fields.map(f=>'${args.'+f.name+'}').join(', ');
  let loadArgs = `{${f.fields.map(f=>`${f.name}`).join(', ')}}`;
  let withArgsTypeof = f.fields.map(f=>'${'+`typeof args.${f.name}`+'}').join(', ');
    #>
  public async findOneBy#{findBy}AndUpdate(#{findArgs}, payload: any) {
    logger.trace(`findOneBy#{findBy}AndUpdate with #{withArgs} `);
    let entity = this.getPayload(payload, true);
    let result = await this.loaders.by#{findBy}.load(#{loadArgs});
    if(result){
      for (let f in entity) {
        if (entity.hasOwnProperty(f)) {
          result.set(f, entity[f]);
        }
      }
      result = await result.save();
      this.storeToCache([result]);
      return this.ensureId(result ? result.toJSON() : result);
    } else {
      return result;
    }
  }
<#-});#>

<#- for (let f of entity.args.remove) {
  let ukey = f.name;
  let type = f.type;
  #>

  public async findOneBy#{f.cName}AndRemove(#{ukey}: #{type}) {
    logger.trace(`findOneBy#{f.cName}AndRemove`);
    let result = await this.loaders.by#{f.cName}.load(#{ukey});
    if( result ){
      result = await result.remove();
      this.storeToCache([result]);
      return this.ensureId(result ? result.toJSON() : result);
    } else {
      return result;
    }
  }
<#-}#>

<#- entity.complexUniqueIndex.forEach(f=> {
  let findBy = f.fields.map(f=>f.uName).join('And');
  let findArgs = f.fields.map(f=>`${f.name}: ${f.type}`).join(', ');
  let withArgs = f.fields.map(f=>'${'+f.name+'}').join(', ');
  let withArgs2 = f.fields.map(f=>'${args.'+f.name+'}').join(', ');
  let loadArgs = `{${f.fields.map(f=>`${f.name}`).join(', ')}}`;
  let withArgsTypeof = f.fields.map(f=>'${'+`typeof args.${f.name}`+'}').join(', ');
    #>
  public async findOneBy#{findBy}AndRemove(#{findArgs}) {
    logger.trace(`findOneBy#{findBy}AndRemove with #{withArgs} `);
    let result = await this.loaders.by#{findBy}.load(#{loadArgs});
    if( result ){
      result = await result.remove();
      this.storeToCache([result]);
      return this.ensureId(result ? result.toJSON() : result);
    } else {
      return result;
    }
  }
<#-});#>

<#- for (let connection of entity.relations) {#>
  public async addTo#{ connection.shortName }(args: {
    <#- for (let f of connection.addArgs) {#>
      #{f.name}?: #{f.type},
    <#-}#>
  }) {
    logger.trace(`addTo#{ connection.shortName }`);
<#- if (connection.verb === 'HasOne') {#>
    let current = await this.findOneById(args.#{entity.ownerFieldName});
    if (current) {
      await this.connectors.#{connection.ref.entity}.findOneByIdAndUpdate(
        args.#{connection.refFieldName},
        { #{connection.ref.field}: current.#{connection.ref.backField} });
    }
<#} else if (connection.verb === 'HasMany') {#>
    let current = await this.findOneById(args.#{entity.ownerFieldName});
    if (current) {
      await this.connectors.#{connection.ref.entity}.findOneByIdAndUpdate(
        args.#{connection.refFieldName},
        { #{connection.ref.field}: current.#{connection.ref.backField}});
    }
<#} else if (connection.verb === 'BelongsTo') {#>
    let opposite = await this.connectors.#{connection.ref.entity}.findOneById(args.#{connection.refFieldName} );
    if (opposite) {
      await this.findOneByIdAndUpdate(args.#{entity.ownerFieldName},
      { #{connection.field}: opposite.#{connection.ref.field} });
    }
<#} else if (connection.verb === 'BelongsToMany') {#>
    let current = await this.findOneById(args.#{entity.ownerFieldName});
    let opposite = await this.connectors.#{connection.ref.entity}.findOneById(args.#{connection.refFieldName} );
    if (current && opposite) {
      let update: any = {
        #{connection.ref.using.field}: current.#{connection.ref.backField},
        #{connection.ref.usingField}: opposite.#{connection.ref.field},
      };
<# for (let fname of connection.ref.fields) {
      if (fname !== 'id') { #>
      if (args.hasOwnProperty('#{fname}')) {
        update.#{fname} = args.#{fname};
      }
<#- }
  }-#>

      let connection = await this.connectors.#{connection.ref.using.entity}.getList({
        #{connection.ref.using.field}: current.#{connection.ref.backField},
        #{connection.ref.usingField}: opposite.#{connection.ref.field},
      });

      if (connection.length > 0) {
        await this.connectors.#{connection.ref.using.entity}.findOneByIdAndUpdate(connection[0]._id, update);
      } else {
        await this.connectors.#{connection.ref.using.entity}.create(update);
      }
    }
<#}-#>
  }

  public async removeFrom#{ connection.shortName }(args: {
    <#- for (let f of connection.removeArgs) {#>
      #{f.name}?: #{f.type},
    <#-}#>
  }) {
    logger.trace(`removeFrom#{ connection.shortName }`);
<#- if (connection.verb === 'HasOne') {#>
    await this.connectors.#{connection.ref.entity}.findOneByIdAndUpdate(args.#{connection.refFieldName},
    { #{connection.ref.field}: null });
<#} else if (connection.verb === 'HasMany') {#>
    await this.connectors.#{connection.ref.entity}.findOneByIdAndUpdate(args.#{connection.refFieldName},
    { #{connection.ref.field}: null });
<#} else if (connection.verb === 'BelongsTo') {#>
    await this.findOneByIdAndUpdate(args.#{entity.ownerFieldName}, { #{connection.field}: null });
<#} else if (connection.verb === 'BelongsToMany') {#>
    let current = await this.findOneById(args.#{entity.ownerFieldName});
    let opposite = await this.connectors.#{connection.ref.entity}.findOneById( args.#{connection.refFieldName} );
    if (current && opposite) {
      let connection = await this.connectors.#{connection.ref.using.entity}.getList({
        #{connection.ref.using.field}: current.#{connection.ref.backField},
        #{connection.ref.usingField}: opposite.#{connection.ref.field},
      });

      if (connection.length > 0) {
        await this.connectors.#{connection.ref.using.entity}.findOneByIdAndRemove(connection[0]._id);
      }
    }
<#}-#>
  }
<#}-#>

<#- for (let f of entity.args.getOne) {
  let ukey = f.name;
  let type = f.type;
    #>
  public async findOneBy#{f.cName}(#{ukey}?: #{type}) {
    logger.trace(`findOneBy#{f.cName} with ${#{ukey}} `);
    let result = await this.loaders.by#{f.cName}.load(#{ukey});
    return this.ensureId(result ? result.toJSON() : result);
  }

  public async _findOneBy#{f.cName}(#{ukey}?: #{type}) {
    logger.trace(`_findOneBy#{f.cName} with ${#{ukey}} ${typeof #{ukey}} `);
    let condition: any;
    if (
    <#- if(type == 'string'){ #>
        #{ukey} !== undefined
        && #{ukey} !== ''
    <#- } else if(type == 'number') {#>
        #{ukey} !== undefined
        && !isNaN(Number(#{ukey}))
    <#- } else if(type == 'boolean'){#>
        #{ukey} !== undefined
        && #{ukey} !== ''
        && (!!args.#{ukey}.match(/true/i) || !!args.#{ukey}.match(/false/i))
    <#}-#>
    ) {
      condition = {#{ukey}};
    }
    if (!condition) {
      logger.error('no unique key provided findOneBy#{f.cName} "#{entity.name}"');
      throw new Error('no unique key provided findOneBy#{f.cName} "#{entity.name}"');
    }
    <#-if (ukey === 'id') {#>
    let result = await this.model.findById(condition.id).exec();
    <#-} else {#>
    let result = await this.model.findOne(condition).exec();
    <#-}#>
    return result;
  }

<#-}-#>

<#- entity.complexUniqueIndex.forEach(f=> {
  let findBy = f.fields.map(f=>f.uName).join('And');
  let findArgs = f.fields.map(f=>`${f.name}: ${f.type}`).join(', ');
  let withArgs = f.fields.map(f=>'${'+f.name+'}').join(', ');
  let withArgs2 = f.fields.map(f=>'${args.'+f.name+'}').join(', ');
  let loadArgs = `{${f.fields.map(f=>`${f.name}`).join(', ')}}`;
  let withArgsTypeof = f.fields.map(f=>'${'+`typeof args.${f.name}`+'}').join(', ');
    #>
  public async findOneBy#{findBy}(#{findArgs}) {
    logger.trace(`findOneBy#{findBy} with #{withArgs} `);
    let result = await this.loaders.by#{findBy}.load(#{loadArgs});
    return this.ensureId(result ? result.toJSON() : result);
  }

  public async _findOneBy#{findBy}(args: {#{findArgs}}) {
    logger.trace(`_findOneBy#{findBy} with #{withArgs2} #{withArgsTypeof} `);
    let condition: any = {};
<# for (let i=0, len = f.fields.length; i < len; i++ ){
  let ukey = f.fields[i].name;
  let type = f.fields[i].type;
  #>
    if (
    <#- if(type == 'string') { #>
        args.#{ukey} !== undefined
        && args.#{ukey} !== ''
    <#- } else if(type == 'number') {#>
        args.#{ukey} !== undefined
        && !isNaN(Number(args.#{ukey}))
    <#- } else if(type == 'boolean'){#>
       args.#{ukey} !== undefined
        && args.#{ukey} !== ''
        && (!!args.#{ukey}.match(/true/i) || !!args.#{ukey}.match(/false/i))
    <#}-#>
    ) {
      condition = {
        ...condition,
        #{ukey}: args.#{ukey},
        };
    } else {
      throw new Error('All parameters required for findOneBy#{findBy}! "#{entity.name}"');
    }
<#}#>
    let result = await this.model.findOne(condition).exec();

    return result;
  }

<#-});-#>


  public getPayload(args: {
    <#- for (let f of entity.args.create) {#>
      #{f.name}?: #{f.type},
    <#-}#>
  }, update?: boolean) {
    let entity: any = {};
    <#- for (let f of entity.args.create) {#>
      if (args.#{f.name} !== undefined) {
        entity.#{f.name} = args.#{f.name};
      }
    <#-}#>
    if (update) {
      delete entity.id;
      delete entity._id;
    } else {
      if (entity.id) {
        entity._id = entity.id;
        delete entity.id;
      }
    }
    return entity;
  }

  public getSort(args) {
    let result: any = {};
<#- if(entity.filterAndSort.length > 0){ -#>
    if (Array.isArray(args.orderBy) && args.orderBy.length > 0) {
      for (let i = 0, len = args.orderBy.length; i < len; i++) {
      let orderBy = args.orderBy[i];
      // tslint:disable-next-line:switch-default
        switch (orderBy) {
  <#- for( let f of entity.filterAndSort){ #>
          case '#{f.name}Asc':
            result.#{f.name} = 1 ;
            break;
          case '#{f.name}Desc':
            result.#{f.name} = -1;
            break;
  <#- } -#>
        }
      }
    }
<# } #>
    // must be last
    result._id = 1;
    return result;
  }

  public  getFilter(args) {
    let result: any = {};
<#- if(entity.filterAndSort.length > 0){ #>
    if (args.filter) {
      result = { $or: [ ] };
<# for( let i = 0, len = entity.filterAndSort.length; i < len; i++) {
  let f = entity.filterAndSort[i]; #>
<#- if(f.type == 'string' && f.gqlType != 'ID'){-#>
      result.$or.push({ #{f.name}: { $regex: `^${args.filter}`, $options: 'im' }});
<#- } else if(f.type == 'string' && f.gqlType == 'ID'){-#>
      result.$or.push({ #{f.name}: args.filter});
<#- } else if (f.type == 'number') {-#>
      if (!isNaN(Number(args.filter))) {
        result.$or.push({ #{f.name}: { $eq: Number(args.filter)}});
      }
<#- } else if (f.type == 'boolean'){-#>
      if ((!!args.filter.match(/true/i) || !!args.filter.match(/false/i))) {
        result.$or.push({ #{f.name}: { $eq: !!args.filter.match(/true/i)}});
      }
<#- }#>
<#}-#>
    }
<#- } #>

<#- for( let f of entity.search){ #>
    if (args.#{f.name} !== undefined
      && args.#{f.name} !== null
      && typeof args.#{f.name} === 'object'
      && args.#{f.name}.constructor === Object) {
      result.#{f._name || f.name} = args.#{f.name};
    } else
<#- if(f.type == 'string' && !(f.gqlType == 'ID' || f.rel) ){ -#>  if (args.#{f.name} !== undefined && args.#{f.name} !== '') {
      result.#{f.name} = { $regex: `^${args.#{f.name}}`, $options: 'im' };
    }
<#- } else if(f.type == 'string' && (f.gqlType == 'ID' || f.rel)){-#> if (args.#{f.name} !== undefined && args.#{f.name} !== '') {
      result.#{f.name} = args.#{f.name};
    }
<#- } else if(f.type == 'number') {-#>  if (args.#{f.name} !== undefined && !isNaN(Number(args.#{f.name}))) {
      result.#{f.name} = { $eq: Number(args.#{f.name})};
    }
<#- } else if(f.type == 'boolean'){-#>  if (args.#{f.name} !== undefined ) {
      result.#{f.name} = { $eq: !!args.#{f.name} };
    }
<#}-#>
<#}#>
    return result;
  }
};
