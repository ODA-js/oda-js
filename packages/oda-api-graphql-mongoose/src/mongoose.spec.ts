import 'jest';
import { FilterMongoose as Filter } from './mongoose';

describe('filter', () => {
  it('unfold queries', () => {
    const result = Filter.parse({ location: { query: { type: 'Point' } } });
    expect(result).toMatchSnapshot();
    const result2 = Filter.parse({
      location: { query: { or: [{ type: 'Point' }, { type: 'Line' }] } },
    });
    expect(result2).toMatchSnapshot();
    const result3 = Filter.parse({
      location: {
        query: {
          or: [
            { type: 'Point' },
            { type: 'Line' },
            {
              and: [
                { coordinates: { in: [40, 50] } },
                { coordinates: { in: [20, 30] } },
              ],
            },
          ],
        },
      },
    });
    expect(result3).toMatchSnapshot();

    const result4 = Filter.parse({
      location: {
        query: {
          and: [
            { type: { eq: 'Point' } },
            { coordinates: { at_0: { gte: 4, lte: 50 } } },
          ],
        },
      },
    });
    expect(result4).toMatchSnapshot();

    const result5 = Filter.parse({
      location: {
        query: {
          coordinates: { all: [40, 5] },
        },
      },
    });
    expect(result5).toMatchSnapshot();

    const result6 = Filter.parse({
      location: {
        query: {
          coordinates: { size: 2 },
        },
      },
    });
    expect(result6).toMatchSnapshot();
    const result7 = Filter.parse({
      location: {
        query: {
          near: {
            geometry: {
              type: 'Point',
              coordinates: [70, 5],
            },
            maxDistance: 50,
            minDistance: 0,
          },
        },
      },
    });
    expect(result7).toMatchSnapshot();
  });
});
