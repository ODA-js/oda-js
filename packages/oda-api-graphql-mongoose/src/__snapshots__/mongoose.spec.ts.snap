// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filter unfold queries 1`] = `
Object {
  "location.type": Object {
    "$eq": "Point",
  },
}
`;

exports[`filter unfold queries 2`] = `
Object {
  "$or": Array [
    Object {
      "location.type": Object {
        "$eq": "Point",
      },
    },
    Object {
      "location.type": Object {
        "$eq": "Line",
      },
    },
  ],
}
`;

exports[`filter unfold queries 3`] = `
Object {
  "$or": Array [
    Object {
      "location.type": Object {
        "$eq": "Point",
      },
    },
    Object {
      "location.type": Object {
        "$eq": "Line",
      },
    },
    Object {
      "$and": Array [
        Object {
          "location.coordinates": Object {
            "$in": Array [
              40,
              50,
            ],
          },
        },
        Object {
          "location.coordinates": Object {
            "$in": Array [
              20,
              30,
            ],
          },
        },
      ],
    },
  ],
}
`;

exports[`filter unfold queries 4`] = `
Object {
  "$and": Array [
    Object {
      "location.type": Object {
        "$eq": "Point",
      },
    },
    Object {
      "location.coordinates.0": Object {
        "$gte": 4,
        "$lte": 50,
      },
    },
  ],
}
`;

exports[`filter unfold queries 5`] = `
Object {
  "location.coordinates": Object {
    "$all": Array [
      40,
      5,
    ],
  },
}
`;

exports[`filter unfold queries 6`] = `
Object {
  "location.coordinates": Object {
    "$size": 2,
  },
}
`;

exports[`filter unfold queries 7`] = `
Object {
  "location": Object {
    "$near": Object {
      "$geometry": Object {
        "coordinates": Array [
          70,
          5,
        ],
        "type": "Point",
      },
      "$maxDistance": 50,
      "$minDistance": 0,
    },
  },
}
`;
